dlib/global_optimization/find_max_global.h

```diff
@@ -949,10 +995,11 @@ template <typename T> static auto go(T&& f, const matrix<double, 0, 1>& a) -> de
         const std::vector<bool>& is_integer_variable,
         const std::chrono::nanoseconds max_runtime,
         double solver_epsilon = 0,
-        const std::vector<function_evaluation>& initial_function_evals = {}
+        const std::vector<function_evaluation>& initial_function_evals = {},
+        stop_condition should_stop = default_stop_condition
```

Comment:
So instead of adding more parameters here it would be better to do this:
```
    template <
        typename funct,
        typename... Args
        >
    function_evaluation find_min_global (
        thread_pool& tp, 
        funct f,
        const matrix<double,0,1>& bound1,
        const matrix<double,0,1>& bound2,
        const std::vector<bool>& is_integer_variable,
        const std::chrono::nanoseconds max_runtime,
        Args&&... args
    )   
    {   
        return find_min_global(tp, std::move(f), bound1, bound2, is_integer_variable, max_function_calls(), max_runtime, solver_epsilon, std::forward<Args>(args)...);
    } 
```

Then the function that actually implements all this can take these things as parameters and the `stop_condition` can just be a `std::function`.  I would otherwise suggest a universal reference instead of a `std::function` but if someone wants to add more arguments to this function later it will work out better if the type is something more narrow than universal reference.

Reply:
How broadly do you want this applied?  There are 20+ overloads of either find_{min,max}_global each with subtly different arguments, and replacing these calls with a variadic template that maintains all of the old usages could be tricky depending on how many I need to move to the variadic template.  I also don't know how well default arguments work for function parameters that have template type on older compilers like gcc 4.X.

Reply:
It will only work for the two arguments I suggested: initial_function_evals and should_stop. So do it only for those two. But do it to all the overloads.

Reply:
I just finished trying this, and unfortunately this would be a breaking change in the API to apply this change to all of the overloads, so I applied the change to as many as I could without breaking the API.  This allowed me to update all the calls that don't invoke `impl::find_max_global`directly.

Reply:
What part would be a breaking change?  I just looked over the code and I don't see it.  Like if `impl::find_max_global()` is the only place that declares `initial_function_evals` and `should_stop` it sure looks like using a parameter pack in all other locations would work and and not break the API.

Reply:
The compile fails when building the python bindings for some reason when passing an internalizer list for initial_function_evals to impl::find_max_global when using perfect forwarding because the compliler deduces an incomplete parameter pack and discards correct the template candidate.  I'm guessing we hit some obscure portion of the standard relating to perfect forwarding and defaulted arguments.

Here are the relevant portions of the compiler error

```console
/home/runderwood/git/contrib/dlib/dlib/../dlib/global_optimization/find_max_global.h:378:16: error: no matching function for call to 'find_max_global'
        return find_max_global(functions, std::move(specs), num, max_runtime, solver_epsilon, {initial_function_evals}).second;
               ^~~~~~~~~~~~~~~
/home/runderwood/git/contrib/dlib/dlib/../dlib/global_optimization/find_max_global.h:460:16: note: in instantiation of function template specialization 'dlib::find_max_global<(lambda at /home/runderwood/git/contrib/dlib/tools/python/src/global_optimization.cpp:124:17)>' requested here
        return find_max_global(std::move(f), bound1, bound2, is_integer_variable, num, FOREVER, solver_epsilon, std::forward<Args>(args)...);
               ^
/home/runderwood/git/contrib/dlib/tools/python/src/global_optimization.cpp:129:19: note: in instantiation of function template specialization 'dlib::find_max_global<(lambda at /home/runderwood/git/contrib/dlib/tools/python/src/global_optimization.cpp:124:17)>' requested here
    auto result = find_max_global(func, list_to_mat(bound1), list_to_mat(bound2),
                  ^
[...snip...]

/home/runderwood/git/contrib/dlib/dlib/../dlib/global_optimization/find_max_global.h:297:43: note: candidate template ignored: substitution failure [with funct = (lambda at /home/runderwood/git/contrib/dlib/tools/python/src/global_optimization.cpp:124:17)]: deduced incomplete pack <(no value)> for template parameter 'Args'
    std::pair<size_t,function_evaluation> find_max_global (
                                          ^

```

Reply:
Oh yeah, duh.  I wasn't paying attention.  It's because some of the `initial_function_evals` are `std::vector<function_evaluation>` while others are `std::vector<std::vector<function_evaluation>>`.   Let me see if I can refactor this a little bit so there aren't so many overloads.  It's kinda gotten out of hand.  I would have done that to begin with but I didn't want to conflict your pull request.  Anyway, I'll see if I can simplify this.

Reply:
I just pushed a change that simplifies this a lot.  So you should now only have to add a parameter to the main top function and all the overloads will just automatically incorporate it.